numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -962,16 +963,21 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
 
 #define PyArray_MAX(a,b) (((a)>(b))?(a):(b))
 #define PyArray_MIN(a,b) (((a)<(b))?(a):(b))
-#define PyArray_CLT(p,q) ((((p).real==(q).real) ? ((p).imag < (q).imag) : \
-                               ((p).real < (q).real)))
-#define PyArray_CGT(p,q) ((((p).real==(q).real) ? ((p).imag > (q).imag) : \
-                               ((p).real > (q).real)))
-#define PyArray_CLE(p,q) ((((p).real==(q).real) ? ((p).imag <= (q).imag) : \
-                               ((p).real <= (q).real)))
-#define PyArray_CGE(p,q) ((((p).real==(q).real) ? ((p).imag >= (q).imag) : \
-                               ((p).real >= (q).real)))
-#define PyArray_CEQ(p,q) (((p).real==(q).real) && ((p).imag == (q).imag))
-#define PyArray_CNE(p,q) (((p).real!=(q).real) || ((p).imag != (q).imag))
+
+#define PyArray_CLT(p,q, type) (NPY_##type##_GET_REAL(&(p)) == NPY_##type##_GET_REAL(&(q)) \
+                                ? NPY_##type##_GET_IMAG(&(p)) < NPY_##type##_GET_IMAG(&(q)) \
+                                : NPY_##type##_GET_REAL(&(p)) < NPY_##type##_GET_REAL(&(q)))
+#define PyArray_CGT(p,q, type) (NPY_##type##_GET_REAL(&(p)) == NPY_##type##_GET_REAL(&(q)) \
+                                ? NPY_##type##_GET_IMAG(&(p)) > NPY_##type##_GET_IMAG(&(q)) \
+                                : NPY_##type##_GET_REAL(&(p)) > NPY_##type##_GET_REAL(&(q)))
+#define PyArray_CLE(p,q, type) (NPY_##type##_GET_REAL(&(p)) == NPY_##type##_GET_REAL(&(q)) \
+                                ? NPY_##type##_GET_IMAG(&(p)) <= NPY_##type##_GET_IMAG(&(q)) \
+                                : NPY_##type##_GET_REAL(&(p)) <= NPY_##type##_GET_REAL(&(q)))
+#define PyArray_CGE(p,q, type) (NPY_##type##_GET_REAL(&(p)) == NPY_##type##_GET_REAL(&(q)) \
+                                ? NPY_##type##_GET_IMAG(&(p)) >= NPY_##type##_GET_IMAG(&(q)) \
+                                : NPY_##type##_GET_REAL(&(p)) >= NPY_##type##_GET_REAL(&(q)))
+#define PyArray_CEQ(p,q, type) ((p) == (q))
+#define PyArray_CNE(p,q, type) ((p) != (q))
```

Comment:
Can we move all these to `npy_math.h` (including `PyArray_MAX`, `PyArray_MIN`)? I don't see use of them in the C code of the [top 4000 packages](https://github.com/search?q=repo%3Ahpyproject%2Ftop4000-pypi-packages+PyArray_C+NOT+path%3A%2F%5Etop100%5C%2F0018-numpy-1.20.3%5C%2Fnumpy%5C%2F%2F&type=code&p=3). I don't see any use of the `PyArray_C*` macros internally either. It is too bad we cannot drop them for NumPy2.0

Reply:
> It is too bad we cannot drop them for NumPy2.0

if they're not useful: why not? these look like terrible names and every user of the C API has to rebuild anyway, so 2.0 is a good place to clean up. or does it interfere with 2.0/1.X compat if they go?

Reply:
> does it interfere with 2.0/1.X compat if they go?

@seberg: thoughts?

Reply:
I agree, almost no users.  Removing API is explicitly OK (in this case it is even possible without forcing to rebuild).

The only thing would be complains of "it didn't go through a deprecation", and since there are no known users and it is easy to vendor (we do it ourselves!) I think that is fine.

Reply:
Will remove them then.

Reply:
Just opened #24258 so that the changes do not get lost in this PR. I'll open an issue and link the PR to that issue. Also, already removed the macros from the docs and will add the note to the upcoming changes. Thanks for the help all!